File: core/models/dto.py
"""
Application DTO model definitions.

Domain/validation layer for :class:`core.models.db.Project`. Nothing here
is persisted directly — the JSONB columns on the ORM model are only ever
written via ``ProjectFeatures.model_dump(mode="json")`` and read back via
``ProjectFeatures.model_validate(...)``, so a raw dict never reaches the
database and a raw dict never leaks back out to callers.

The vocabulary here is shared by both ``ProjectType.CONDOMINIUM`` and
``ProjectType.VILLA`` projects. Project type determines which fields are
meaningful, not what the fields are called — see
``core/models/project_entity_guide.md`` for the full rationale. A "private
swimming pool" is ``facilities.pool.private``, not a separate
``private_swimming_pool`` field; a "rooftop lounge" is ``lounge.rooftop``,
not a separate ``rooftop_lounge`` field; and so on for every qualifier
below.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from enum import StrEnum

from pydantic import BaseModel, Field


class ProjectType(StrEnum):
    """Type of property project."""

    CONDOMINIUM = "condominium"
    VILLA = "villa"


class ProjectStatus(StrEnum):
    """Lifecycle status of a project."""

    PLANNED = "planned"
    UNDER_CONSTRUCTION = "under_construction"
    COMPLETED = "completed"


class ProjectLocation(BaseModel):
    """Structured location for a project (never a free-text string)."""

    address: str | None = None
    district: str | None = None
    city: str | None = None
    country: str
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)


# --- Facilities: things provided by the project/development ---------------


class PoolFeatures(BaseModel):
    available: bool = False
    private: bool = False
    children: bool = False
    deck: bool = False
    shower: bool = False


class WellnessFeatures(BaseModel):
    sauna: bool = False
    spa: bool = False


class SecurityFeatures(BaseModel):
    security_24_7: bool = False
    cctv: bool = False


class AccessFeatures(BaseModel):
    key_card: bool = False
    lobby_reception: bool = False


class ParkingFeatures(BaseModel):
    available: bool = False
    garage: bool = False


class LoungeFeatures(BaseModel):
    """Canonical lounge concept, shared by facilities and outdoor spaces.

    Replaces the would-be duplicate fields ``lounge``, ``outdoor_lounge``
    and ``rooftop_lounge`` with qualifiers on a single object.
    """

    available: bool = False
    indoor: bool = False
    outdoor: bool = False
    rooftop: bool = False


class Facilities(BaseModel):
    """Things provided by the project/development."""

    pool: PoolFeatures = Field(default_factory=PoolFeatures)
    fitness_center: bool = False
    wellness: WellnessFeatures = Field(default_factory=WellnessFeatures)

    kids_club: bool = False

    cinema: bool = False
    library: bool = False
    gaming_room: bool = False

    lounge: LoungeFeatures = Field(default_factory=LoungeFeatures)

    security: SecurityFeatures = Field(default_factory=SecurityFeatures)
    access: AccessFeatures = Field(default_factory=AccessFeatures)
    parking: ParkingFeatures = Field(default_factory=ParkingFeatures)


# --- Spaces: physical spaces rather than services --------------------------


class InteriorSpaces(BaseModel):
    living_room: bool = False
    dining_area: bool = False
    kitchen: bool = False
    master_bedroom: bool = False
    guest_bedrooms: int | None = Field(default=None, ge=0)
    bathrooms: int | None = Field(default=None, ge=0)
    walk_in_wardrobe: bool = False
    laundry: bool = False
    storage: bool = False
    staff_room: bool = False
    guest_house: bool = False


class GardenFeatures(BaseModel):
    available: bool = False
    private: bool = False
    tropical: bool = False


class TerraceFeatures(BaseModel):
    available: bool = False
    covered: bool = False
    rooftop: bool = False
    sea_view: bool = False


class OutdoorSpaces(BaseModel):
    garden: GardenFeatures = Field(default_factory=GardenFeatures)
    pool_area: bool = False
    terrace: TerraceFeatures = Field(default_factory=TerraceFeatures)
    lounge: LoungeFeatures = Field(default_factory=LoungeFeatures)

    outdoor_dining: bool = False
    outdoor_kitchen: bool = False

    sala: bool = False
    walking_paths: bool = False
    children_area: bool = False


class Spaces(BaseModel):
    interior: InteriorSpaces = Field(default_factory=InteriorSpaces)
    outdoor: OutdoorSpaces = Field(default_factory=OutdoorSpaces)


# --- Features: characteristics that aren't a physical "space" --------------


class Features(BaseModel):
    air_conditioning: bool = False
    smart_home: bool = False
    bbq: bool = False
    jacuzzi: bool = False
    wine_cellar: bool = False
    automatic_gate: bool = False

    kitchen_appliances: list[str] = Field(default_factory=list)


# --- Nearby places -----------------------------------------------------


class NearbyPlace(BaseModel):
    name: str
    distance_m: int | None = Field(default=None, ge=0)


class NearbyPlaces(BaseModel):
    airports: list[NearbyPlace] = Field(default_factory=list)
    international_schools: list[NearbyPlace] = Field(default_factory=list)
    shopping_centers: list[NearbyPlace] = Field(default_factory=list)
    golf_courses: list[NearbyPlace] = Field(default_factory=list)
    marinas: list[NearbyPlace] = Field(default_factory=list)
    tourist_attractions: list[NearbyPlace] = Field(default_factory=list)


# --- Aggregate ---------------------------------------------------------


class ProjectFeatures(BaseModel):
    """The full contents of ``Project.features`` (a single JSONB column)."""

    facilities: Facilities = Field(default_factory=Facilities)
    spaces: Spaces = Field(default_factory=Spaces)
    features: Features = Field(default_factory=Features)
    nearby: NearbyPlaces = Field(default_factory=NearbyPlaces)


class ProjectCreate(BaseModel):
    """Application-level payload for creating a :class:`~core.models.db.Project`."""

    project_type: ProjectType
    name: str
    status: ProjectStatus = ProjectStatus.PLANNED

    location: ProjectLocation

    distance_to_beach_m: int | None = Field(default=None, ge=0)
    total_units: int | None = Field(default=None, ge=0)
    total_buildings: int | None = Field(default=None, ge=0)
    completion_date: date | None = None

    features: ProjectFeatures = Field(default_factory=ProjectFeatures)


class ProjectUpdate(ProjectCreate):
    """Application-level payload for updating a :class:`~core.models.db.Project`.

    All fields are optional, since the user may only want to update a subset
    of them.
    """

    project_type: ProjectType | None = None
    name: str | None = None
    status: ProjectStatus | None = None

    location: ProjectLocation | None = None

    distance_to_beach_m: int | None = Field(default=None, ge=0)
    total_units: int | None = Field(default=None, ge=0)
    total_buildings: int | None = Field(default=None, ge=0)
    completion_date: date | None = None

    features: ProjectFeatures | None = None


class ProjectResponse(ProjectCreate):
    """Response model for a :class:`~core.models.db.Project`."""

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class BlogStatus(StrEnum):
    """Lifecycle status of a blog post."""

    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


class BlogCreate(BaseModel):
    """Application-level payload for creating a :class:`~core.models.db.Blog`."""

    title: str
    slug: str
    excerpt: str | None = None
    content: str
    featured_image_url: str | None = None

    status: BlogStatus = BlogStatus.DRAFT

    author_id: uuid.UUID
    category_id: uuid.UUID | None = None

    published_at: datetime | None = None

    seo_title: str | None = None
    seo_description: str | None = None


class BlogUpdate(BlogCreate):
    """Application-level payload for updating a :class:`~core.models.db.Blog`.

    All fields are optional, since the user may only want to update a subset
    of them.
    """

    title: str | None = None
    slug: str | None = None
    excerpt: str | None = None
    content: str | None = None
    featured_image_url: str | None = None

    status: BlogStatus | None = None

    author_id: uuid.UUID | None = None
    category_id: uuid.UUID | None = None

    published_at: datetime | None = None

    seo_title: str | None = None
    seo_description: str | None = None


class BlogResponse(BlogCreate):
    """Response model for a :class:`~core.models.db.Blog`."""

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class PaginationMetadata(BaseModel):
    """Describes the pagination state of a list / search response.

    Attributes:
        total_items:  Total number of matching records in the database.
        total_pages:  Total number of pages given the current ``limit``.
        current_page: 1-based page number derived from ``skip`` and ``limit``.
        skip:         The offset that was applied.
        limit:        The maximum number of items per page.
    """

    total_items: int = Field(..., description="Total number of matching records.")
    total_pages: int = Field(..., description="Total number of pages.")
    current_page: int = Field(..., description="Current page number (1-based).")
    skip: int = Field(..., description="Number of records skipped.")
    limit: int = Field(..., description="Maximum records per page.")


class PagninatedResponse[T: BaseModel](BaseModel):
    """Generic paginated response envelope.

    Type parameter ``T`` is the item model (e.g. ``SellerResponse``).

    Example JSON::

        {
          "metadata": { ... },
          "items": [ ... ]
        }
    """

    metadata: PaginationMetadata = Field(..., description="Pagination metadata.")
    items: list[T] = Field(..., description="Page of result items.")

File: web/ui/src/services/dto.ts
/**
 * The wire format of `web/api`, mirrored field for field from
 * `core/models/dto.py`.
 *
 * These are the shapes as they arrive — snake_case, nullable where the
 * Pydantic model is `| None`. Nothing outside `mappers.ts` should import from
 * here; the rest of the site works in the camelCase view models declared in
 * `src/data/`. That indirection is the whole point: a rename on the Python
 * side is a one-file change here plus a one-file change in the mapper.
 */

// --- Enums (Python StrEnum, so they arrive as plain strings) ---------------

export type ProjectTypeDto = "condominium" | "villa";
export type ProjectStatusDto = "planned" | "under_construction" | "completed";
export type BlogStatusDto = "draft" | "published" | "archived";

// --- Location --------------------------------------------------------------

export interface ProjectLocationDto {
  address: string | null;
  district: string | null;
  city: string | null;
  country: string;
  latitude: number | null;
  longitude: number | null;
}

// --- Facilities: things provided by the development ------------------------

export interface PoolFeaturesDto {
  available: boolean;
  private: boolean;
  children: boolean;
  deck: boolean;
  shower: boolean;
}

export interface WellnessFeaturesDto {
  sauna: boolean;
  spa: boolean;
}

export interface SecurityFeaturesDto {
  security_24_7: boolean;
  cctv: boolean;
}

export interface AccessFeaturesDto {
  key_card: boolean;
  lobby_reception: boolean;
}

export interface ParkingFeaturesDto {
  available: boolean;
  garage: boolean;
}

/** One lounge concept, qualified — not three separate lounge fields. */
export interface LoungeFeaturesDto {
  available: boolean;
  indoor: boolean;
  outdoor: boolean;
  rooftop: boolean;
}

export interface FacilitiesDto {
  pool: PoolFeaturesDto;
  fitness_center: boolean;
  wellness: WellnessFeaturesDto;
  kids_club: boolean;
  cinema: boolean;
  library: boolean;
  gaming_room: boolean;
  lounge: LoungeFeaturesDto;
  security: SecurityFeaturesDto;
  access: AccessFeaturesDto;
  parking: ParkingFeaturesDto;
}

// --- Spaces ----------------------------------------------------------------

export interface InteriorSpacesDto {
  living_room: boolean;
  dining_area: boolean;
  kitchen: boolean;
  master_bedroom: boolean;
  guest_bedrooms: number | null;
  bathrooms: number | null;
  walk_in_wardrobe: boolean;
  laundry: boolean;
  storage: boolean;
  staff_room: boolean;
  guest_house: boolean;
}

export interface GardenFeaturesDto {
  available: boolean;
  private: boolean;
  tropical: boolean;
}

export interface TerraceFeaturesDto {
  available: boolean;
  covered: boolean;
  rooftop: boolean;
  sea_view: boolean;
}

export interface OutdoorSpacesDto {
  garden: GardenFeaturesDto;
  pool_area: boolean;
  terrace: TerraceFeaturesDto;
  lounge: LoungeFeaturesDto;
  outdoor_dining: boolean;
  outdoor_kitchen: boolean;
  sala: boolean;
  walking_paths: boolean;
  children_area: boolean;
}

export interface SpacesDto {
  interior: InteriorSpacesDto;
  outdoor: OutdoorSpacesDto;
}

// --- Features --------------------------------------------------------------

export interface FeaturesDto {
  air_conditioning: boolean;
  smart_home: boolean;
  bbq: boolean;
  jacuzzi: boolean;
  wine_cellar: boolean;
  automatic_gate: boolean;
  kitchen_appliances: string[];
}

// --- Nearby ----------------------------------------------------------------

export interface NearbyPlaceDto {
  name: string;
  distance_m: number | null;
}

export interface NearbyPlacesDto {
  airports: NearbyPlaceDto[];
  international_schools: NearbyPlaceDto[];
  shopping_centers: NearbyPlaceDto[];
  golf_courses: NearbyPlaceDto[];
  marinas: NearbyPlaceDto[];
  tourist_attractions: NearbyPlaceDto[];
}

/** The contents of the single `features` JSONB column. */
export interface ProjectFeaturesDto {
  facilities: FacilitiesDto;
  spaces: SpacesDto;
  features: FeaturesDto;
  nearby: NearbyPlacesDto;
}

// --- Top-level responses ---------------------------------------------------

export interface ProjectResponseDto {
  id: string;
  project_type: ProjectTypeDto;
  name: string;
  status: ProjectStatusDto;
  location: ProjectLocationDto;
  distance_to_beach_m: number | null;
  total_units: number | null;
  total_buildings: number | null;
  /** ISO date, e.g. "2026-11-30". */
  completion_date: string | null;
  features: ProjectFeaturesDto;
  /** ISO 8601 timestamp. */
  created_at: string;
  /** ISO 8601 timestamp. */
  updated_at: string;
}

export interface BlogResponseDto {
  id: string;
  title: string;
  slug: string;
  excerpt: string | null;
  /** Raw HTML. See the sanitisation note in `web/api/TODO.md`. */
  content: string;
  featured_image_url: string | null;
  status: BlogStatusDto;
  author_id: string;
  category_id: string | null;
  /** ISO 8601 timestamp. */
  published_at: string | null;
  seo_title: string | null;
  seo_description: string | null;
  created_at: string;
  updated_at: string;
}

// --- Pagination envelope ---------------------------------------------------

export interface PaginationMetadataDto {
  total_items: number;
  total_pages: number;
  /** 1-based. */
  current_page: number;
  skip: number;
  limit: number;
}

export interface PaginatedDto<T> {
  metadata: PaginationMetadataDto;
  items: T[];
}

// --- Submissions (planned, see `web/api/TODO.md`) --------------------------

export interface EnquiryCreateDto {
  name: string;
  email: string;
  phone?: string;
  message?: string;
  /** Set when the enquiry came from a property page. */
  project_slug?: string;
}

export interface NewsletterSubscriptionCreateDto {
  email: string;
}

export interface SubmissionResponseDto {
  id: string;
  status: string;
}
